### PR TITLE
Feature/TDKN-92 improve serialization and refs

### PR DESCRIPTION
**Pull request by sgandon**
**Please check if the PR fulfills these requirements**

- [X] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [X] Tests for the changes have been added (for bug fixes / features) ?
- [X] Docs have been added / updated (for bug fixes / features) ?
- [X] The code coverage on new code >75%
- [] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
definition does not provide way to create Properties
json schema serialization misses some information to unserialize
base for handling ReferencesProperties, but this is not used yet cause we don't need to serialize those references.
Serialization did not ignore transient fields.
DefinitionRegistryService did not offer ways to get definitions from a given Properties type 

**What is the new behavior?**
All this above.

**Does this PR introduce a breaking change?**

- [x] Yes
- [] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...
They are addressed in the related PR in *components* and *components-ee*.

**Other information**:

### CODE REVIEW

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/227/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/227/)

File: daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonSchemaGenerator.java [hunk 1]
```diff
@@ -41,11 +43,28 @@ private ObjectNode processTProperties(Properties cProperties) {
         List<Properties> propertiesList = getSubProperties(cProperties);
         for (Properties properties : propertiesList) {
             String name = properties.getName();
-            ((ObjectNode) schema.get(JsonSchemaConstants.TAG_PROPERTIES)).set(name, processTProperties(properties));
+            // if this is a reference then just store it as a string and only store the definition
+            if (properties instanceof ReferenceProperties<?>) {
+                ReferenceProperties<?> referenceProperties = (ReferenceProperties<?>) properties;
+                ((ObjectNode) schema.get(JsonSchemaConstants.TAG_PROPERTIES)).set(name,
```
Comment (bchen-talend) [inline daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonSchemaGenerator.java]: same here, why store reference definition name here?

File: daikon/src/main/java/org/talend/daikon/properties/ReferenceProperties.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        return "reference".equals(f.getName());
+    }
+
+    /**
+     * resolve the referenced properties between a group of properties.
+     * 
+     * @param properties list of all references to resolve
+     * @param definitionRegistry used to find the definitions compatible with current properties
+     */
+    public static void resolveReferenceProperties(final Iterable<? extends Properties> properties,
```
Comment (bchen-talend) [inline daikon/src/main/java/org/talend/daikon/properties/ReferenceProperties.java]: what's the usage of this method?

**Comment by sgandon** [comment]
You are right this is not clear.
ReferenceProperties was an idea to model the fact that some properties may have reference on other properties such as Dataset referencing Datastore. 
The idea behind is that those reference may be serialized if we need to have a widget representing those references, which is not the case for the current dataset and datastore. But also be used when deserializing a set of Properties that are related to each other in order to link back all these reference (resolve them).
For instance when the client asks for the ui-spec for an IO component to the TCOMP service, it will have as inputs the properies (formdata) for a dataset, a datastore, and a IO component. 
```
[ 
    {
      “@definitionName”:”KafkaInput”
      "useCompress": false,
      "configurations": {}
    },
    {
      “@definitionName”:”KafkaDataset”
      "schema": null
    },
    {
      “@definitionName”:”KafkaDatastore”
      "url": null
    }
]
```
The idea is to create each Properties instances, then unserialize them to be filled with the data and a last resolve the ReferenceProperties so that each reference is properly set. And this is done without the knowledge of the link between them. We just use the ReferenceProperties to resolve them.

File: daikon/src/main/java/org/talend/daikon/properties/ReferenceProperties.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+     * the Generic type T for this class must be created by the defition matching the referenceDefintionName.
+     */
+    public final Property<String> referenceDefinitionName = newProperty("referenceDefinitionName");
+
+    /**
+     * the reference instance
+     */
+    private transient T reference;
+
+    private static Iterable<Definition> definitionForPropertyType;
```
Comment (sgandon) [inline daikon/src/main/java/org/talend/daikon/properties/ReferenceProperties.java]: you are right this is a mistake, I'll remove this.

File: daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonDataGenerator.java [hunk 4]
```diff
@@ -29,7 +32,13 @@ private ObjectNode processTPropertiesData(Properties cProperties) {
         List<Properties> propertiesList = getSubProperties(cProperties);
         for (Properties properties : propertiesList) {
             String name = properties.getName();
-            rootNode.set(name, processTPropertiesData(properties));
+            // for ReferencePropertie we just create a String node in the root just like any Property
+            if (properties instanceof ReferenceProperties<?>) {
+                ReferenceProperties<?> referenceProperties = (ReferenceProperties<?>) properties;
+                rootNode.put(properties.getName(), referenceProperties.referenceDefinitionName.getValue());
```
Comment (sgandon) [inline daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonDataGenerator.java]: right this is some code that we first thought was useful for the datastore and dataset but it is not anymore. I have left it here for a future implementation of references with widgets.
Do you think I should remove it ?

File: daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonSchemaGenerator.java (hunk 1 above)
Comment (sgandon) [inline daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonSchemaGenerator.java]: same answer as above.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/240/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/240/)

File: daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonDataGenerator.java (hunk 4 above)
Comment (bchen-talend) [inline daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonDataGenerator.java]: There is a default check for https://mozilla-services.github.io/react-jsonschema-form/, error when there is one property in json schema but do not defined in ui schema. but I do not know if front end use it or not.
So I suggest to remove it and add widget/schema/data together.

File: daikon/src/test/java/org/talend/daikon/serialize/jsonschema/JsonSchemaUtilTest.java [hunk 5]
```diff
@@ -123,40 +120,6 @@ public void testSerializeWithDefinitionName() throws URISyntaxException, IOExcep
         assertEquals(defNameNode.asText(), "ZeDefinitionName");
     }
 
-    @Test
```
Comment (pbailly) [inline daikon/src/test/java/org/talend/daikon/serialize/jsonschema/JsonSchemaUtilTest.java]: This test should not be removed.
There is no justification to remove this test
Plus, it allows us to easily check the JSON generated for a referenced properties.

File: daikon/src/main/java/org/talend/daikon/properties/ReferenceProperties.java [hunk 6]
```diff
@@ -12,40 +12,52 @@
...
+     * the Generic type T for this class must be created by the defition matching the referenceDefintionName.
+     */
+    public final Property<String> referenceDefintionName = newProperty("referenceDefintionName");
+
+    /**
+     * the reference instance
+     */
+    private transient T reference;
+
+    public ReferenceProperties(String name, String referenceDefintionName) {
```
Comment (pbailly) [inline daikon/src/main/java/org/talend/daikon/properties/ReferenceProperties.java]: Could you add a commentary to explicitly describe what is the referenceDefinitionName?

File: daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonSchemaGenerator.java [hunk 7]
```diff
@@ -45,16 +43,28 @@ private ObjectNode processTProperties(Properties cProperties) {
         List<Properties> propertiesList = getSubProperties(cProperties);
         for (Properties properties : propertiesList) {
             String name = properties.getName();
-            if (ReferenceProperties.class.isAssignableFrom(properties.getClass())) {
-                ((ObjectNode) schema.get(JsonSchemaConstants.TAG_PROPERTIES)).put(name,
-                        ((ReferenceProperties) properties).componentType.getValue());
+            // if this is a reference then just store it as a string and only store the definition
+            if (properties instanceof ReferenceProperties<?>) {
```
Comment (pbailly) [inline daikon/src/main/java/org/talend/daikon/serialize/jsonschema/JsonSchemaGenerator.java]: This piece of code will generate JSON schema like:
```
"jsonSchema": {
    "title": "properties.refexample.displayName",
    "type": "object",
    "properties": {
      "parentProp": {
        "title": "property.parentProp.displayName",
        "type": "string"
      },
      "testAPropReference": {
        "title": "properties.testAPropReference.displayName",
        "type": "string"
      }
    }
  }
```
So we are going to display the referenced properties in the UI.
But a referenced property should always be hidden, so we do not want any mention of the ReferenceProperties on this jsonSchema.
This part should removed.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/251/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/251/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/280/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/280/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/283/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/283/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/daikon_pull-request_master/284/) [:star:](https://ci-common.datapwn.com/job/daikon_pull-request_master/284/)

**Pull request merged**